compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -779,6 +779,45 @@ impl Type {
         }
     }
 
+    /// Returns true if a value of this type can safely pass between constrained and
+    /// unconstrained functions (and vice-versa).
+    pub(crate) fn is_valid_for_unconstrained_boundary(&self) -> bool {
+        match self {
+            Type::FieldElement
+            | Type::Integer(_, _)
+            | Type::Bool
+            | Type::Unit
+            | Type::Constant(_)
+            | Type::TypeVariable(_, _)
+            | Type::NamedGeneric(_, _)
+            | Type::Function(_, _, _)
+            | Type::FmtString(_, _)
+            | Type::Error => true,
+
+            Type::Slice(_)
+            | Type::MutableReference(_)
+            | Type::Forall(_, _)
+            | Type::Code
+            | Type::TraitAsType(..) => false,
```

Comment:
Compared to the other "is_valid_for" functions, this is more permissive since it allows FmtStrings and Functions.